dds/idl/be_global.cpp

```diff
@@ -602,3 +613,147 @@ BE_GlobalData::get_include_block(BE_GlobalData::stream_enum_t which)
 
   return ret.str();
 }
+
+void
+BE_GlobalData::cache_annotations()
+{
+  if (idl_global->idl_version_ >= IDL_VERSION_4) {
+    UTL_Scope* root = idl_global->scopes().bottom();
+    topic_annotation_ =
+      AST_Annotation_Decl::narrow_from_decl(
+        root->lookup_by_name("::@topic"));
+    key_annotation_ = AST_Annotation_Decl::narrow_from_decl(
+        root->lookup_by_name("::@key"));
+    nested_annotation_ = AST_Annotation_Decl::narrow_from_decl(
+        root->lookup_by_name("::@nested"));
+    default_nested_annotation_ = AST_Annotation_Decl::narrow_from_decl(
+        root->lookup_by_name("::@default_nested"));
+ }
+}
+
+/**
+ * @brief determines if we should treat a given node as a topic type. This could
+ * mean that it is explicitly a topic type or that @default_nested(FALSE) or
+ * --default_nested = false
+ * @param node the incomming node
+ * @return if type support should be generated
+ * @see BE_GlobalData::is_topic_type(AST_Decl* node)
+ * @see BE_GlobalData::is_default_nested(AST_Decl* node)
+ * @author ceneblock
+ */
+bool BE_GlobalData::treat_as_topic(AST_Decl *node)
+{
+  return is_topic_type(node) || !is_nested_type(node);
+}
+
+/**
+ * @brief determins if a given node is nested by default or not. This may be set
+ * by the module level of @default_nesed or through the --default_nested=
+ * @param node the node to be inspected
+ * @return if the node is default nested or not
+ * @note false means it should be considered nested.
+ * @author ceneblock
+ */
+bool
+BE_GlobalData::is_default_nested(AST_Decl* node)
+{
+  if(node->defined_in()){
+
+    ///recurse to the top
+    bool default_nested_apply_value = is_default_nested(dynamic_cast<AST_Decl *>(node -> defined_in()));
```

Comment:
What is going on with the spaces around "->" ? Also, do these need to be dynamic casts? If so, should we be checking for failure?

Reply:
> What is going on with the spaces around "->" ?

The style for this one is already fixed in changes I haven't pushed.

> Also, do these need to be dynamic casts?

`defined_in` returns an `UTL_Scope`, which is not an `AST_Decl`, but a lot of `AST_Decl`s are `UTL_Scope`s, so yes, the dynamic cast is necessary.

> If so, should we be checking for failure?

If it cast to null it's because the parent node wasn't an `AST_Decl`, which can only mean we hit the root module, which is a case handled in `is_default_nested`.

As for error handling in general, these functions really shouldn't fail; if they do, something is wrong with the fundamental structure of the AST. In addition to that tao_idl is pointer ridden, basically every value that's not a primitive is a pointer. I might I have become numb to it, but I don't want to check every single pointer in a case like this:

```c++
AST_Annotation_Member::narrow_from_decl((*default_nested_apply)["value"])->value ()->ev()->u.bval;
```

If I can be highly certain that these pointers are not going to be null.